### PR TITLE
feat: add RustWasmerWasiRuntime BindingsType

### PR DESCRIPTION
**Pull request by nico151999**
As suggested in #133

### CODE REVIEW

**Review by arendjr — COMMENTED** [review]
Thanks a ton! I haven't looked into the smallest details yet, but I think this approach is solid :+1:

It's good it's a separate runtime for now, so that there's no breaking changes involved. I would like to see some tests before merging, but otherwise I think this could work.

**Comment by nico151999** [comment]
> Thanks a ton! I haven't looked into the smallest details yet, but I think this approach is solid 👍
> 
> It's good it's a separate runtime for now, so that there's no breaking changes involved. I would like to see some tests before merging, but otherwise I think this could work.

Thanks for the response. Currently, there's a lot that I just copy pasted from RustWasmerRuntime. I'm just about to reference the redundant stuff so there are no code duplicates in the final result. Consider the first commit to be the easily readable raw changes and the second commit that will follow shortly to be the cleaner solution.

**Comment by nico151999** [comment]
> > Thanks a ton! I haven't looked into the smallest details yet, but I think this approach is solid 👍
> > It's good it's a separate runtime for now, so that there's no breaking changes involved. I would like to see some tests before merging, but otherwise I think this could work.
> 
> Thanks for the response. Currently, there's a lot that I just copy pasted from RustWasmerRuntime. I'm just about to reference the redundant stuff so there are no code duplicates in the final result. Consider the first commit to be the easily readable raw changes and the second commit that will follow shortly to be the cleaner solution.

@arendjr refactoring is done. Should be ready to merge from my perspective. Would be glad to hear your agreement :)

**Comment by arendjr** [comment]
Very nice! Would you mind adding an "example runtime" with tests, similar to the one in `examples/example-rust-runtime`?

**Comment by nico151999** [comment]
I added RustWasmerWasiRuntime to example-protocol but creating an example runtime like `examples/example-rust-runtime` is another story. `examples/example-rust-runtime` depends on `examples/example-plugin` which is configured for a wasm32-unknown-unknown target. However, we need a wasm32-wasi target in order to create a sound new runtime example for wasi. Any solution you'd prefer?

**Comment by arendjr** [comment]
Good point! I think having an `example-wasi-plugin` would be a very valuable thing to have, so that we can show people how to set up the entire WASI flow. I will happily help out to create some documentation for that as well, and see if I can add an `example-wasi-deno-runtime` as well. Let me know if you need any help with the WASI plugin!

**Comment by nico151999** [comment]
I will try to create a suitable solution tomorrow :)

**Comment by nico151999** [comment]
@arendjr I created a solution. However, it includes some changes you might want to discuss. Would be happy to read your feedback and see this feature released soon :)

File: examples/example-plugin/rust-toolchain.toml
Comment (arendjr) [inline examples/example-plugin/rust-toolchain.toml:5]: Appreciate the explanation!

**Review by arendjr — APPROVED** [review]
Awesome! I'm really happy with this!

Would be nice if we can cut down on some of the duplication, but I don't even consider that a blocker right now.

Do note I have a holiday planned for next week, so I'll probably merge it and prepare the TS runtime after that. Then I think we can prepare a really nice release :+1:

File: examples/example-rust-wasmer-wasi-runtime/src/test.rs
Comment (nico151999) [reply examples/example-rust-wasmer-wasi-runtime/src/test.rs:9]: Basically yes, this is the only major change. If it's a test we should try to eliminate duplicates and use references instead. Maybe we could even create a single rust-wasmer-runtime for both wasm32-unknown-unknown and wasm32-wasi. However, if it's an example meant to serve as a demo it should be easily understandable at the first glance and we might want to have independently working examples a user can read without looking up too many functions' sources in other files. Depending on how we would like to treat it we can act accordingly.

Comment (nico151999) [reply examples/example-rust-wasmer-wasi-runtime/src/test.rs]: I have just done some refactoring which allows for one code basis and does not require code duplicates. However, one needs to pass a parameter to the `cargo test` command to use the WASI related code. I highlighted it in the `examples/README.md` file. The changes were applied in commit [abeeecb](https://github.com/fiberplane/fp-bindgen/pull/135/commits/abeeecb374637ca9733c4d957b374b0f481d82a0).

**Comment by nico151999** [comment]
@arendjr next week I will have exams followed by a hiking trip so I will be off work for some days, too. From my perspective this PR is ready to be merged though. So if you agree we can merge this PR this week already as you suggested.

**Comment by arendjr** [comment]
Thanks so much! Your solution looks nice, but the test command doesn't work for me when I try it locally:

```
$ cargo test -F wasi
error: Found argument '-F' which wasn't expected, or isn't valid in this context

        If you tried to supply `-F` as a value rather than a flag, use `-- -F`

USAGE:
    cargo test [OPTIONS] [TESTNAME] [-- <args>...]

For more information try --help
```

Also tried:
```
cargo test -- -F wasi
    Finished test [unoptimized + debuginfo] target(s) in 0.04s
     Running unittests (/home/arendjr/Projects/nico/target/debug/deps/example_protocol-2692f0ee56c1943a)
error: Unrecognized option: 'F'
error: test failed, to rerun pass '-p example-protocol --bin example-protocol'
```

**Comment by nico151999** [comment]
> Thanks so much! Your solution looks nice, but the test command doesn't work for me when I try it locally:
> 
> ```
> $ cargo test -F wasi
> error: Found argument '-F' which wasn't expected, or isn't valid in this context
> 
>         If you tried to supply `-F` as a value rather than a flag, use `-- -F`
> 
> USAGE:
>     cargo test [OPTIONS] [TESTNAME] [-- <args>...]
> 
> For more information try --help
> ```
> 
> Also tried:
> 
> ```
> cargo test -- -F wasi
>     Finished test [unoptimized + debuginfo] target(s) in 0.04s
>      Running unittests (/home/arendjr/Projects/nico/target/debug/deps/example_protocol-2692f0ee56c1943a)
> error: Unrecognized option: 'F'
> error: test failed, to rerun pass '-p example-protocol --bin example-protocol'
> ```

`cargo test -F wasi` works for me (stable cargo 1.63.0 (fd9c4297c 2022-07-01)). Which version do you use?

**Comment by arendjr** [comment]
Ah, I was still on 1.60 :innocent: It all works now!

**Comment by arendjr** [comment]
Oh, only the CI still fails on what I think is a merge conflict.

**Comment by nico151999** [comment]
> Oh, only the CI still fails on what I think is a merge conflict: [fiberplane/fp-bindgen/runs/7873982677?check_suite_focus=true](https://github.com/fiberplane/fp-bindgen/runs/7873982677?check_suite_focus=true)

Ran the formatter in [ae4d318](https://github.com/fiberplane/fp-bindgen/pull/135/commits/ae4d318ca3a8144ddf1377e1532b2e84cc560a55). Could you try again?

**Comment by arendjr** [comment]
Triggered a new build, but I also noticed some changes in the `expected_bindings.rs` related to a different PR that was merged the other day. You should see it when you run `cargo test` in the root.

**Comment by nico151999** [comment]
> Triggered a new build, but I also noticed some changes in the `expected_bindings.rs` related to a different PR that was merged the other day. You should see it when you run `cargo test` in the root.

Seems like I forgot to update the file. Sorry for the inconvenience. I didn't know which steps need to pass (haven't worked with Github Actions yet but Gitlab pipelines only). Anyway, I can see the jobs that are to be executed now. Will check them.

**Comment by nico151999** [comment]
@arendjr now that I know how Github Actions are configured I also added the wasi test to the last job (or whatever the equivalent in Github Actions is called). Could you try again?

**Comment by arendjr** [comment]
Oh, I think the last failure is because the wasi target doesn't get installed yet. Changing the `rustup target add` line to `rustup target add wasm32-wasi wasm32-unknown-unknown` in the `test.yml` should fix it, I think.

**Comment by nico151999** [comment]
Seems like the wasm32-wasi target is not installed in the pipeline. Will add rustup target add wasm32-wasi to it.

**Comment by nico151999** [comment]
One more thing I'd like to discuss. Do you use commit squashing for merging [PRs](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests)? It would make sense in this case imo.

**Comment by arendjr** [comment]
Yeah, we do :)

**Comment by arendjr** [comment]
Hm, maybe we need to switch to the nightly channel before installing the components?

**Comment by nico151999** [comment]
Sorry for the delayed response. Had to take care about something else. I have just taken a look at it. No need to switch to nightly before adding the targets. I will add a commit right away.

**Comment by arendjr** [comment]
Hm, for some reason GitHub is no longer offering me the "Approve and run" button... Let me see what's up.

**Comment by nico151999** [comment]
I am not 100% sure but the Github web editor found a syntax error that did not make sense to me at first. However, it was gone when I removed the comment in the multiline value of the step's run parameter. Seems like this is not allowed in YAML files even though my local YAML parser did not have any issues parsing it. I hope this was the root cause and it's fixed now. Would you try again, please?

**Comment by arendjr** [comment]
All green :partying_face: 

Thank you so much for this feature and your help with the tests! I'll merge it now indeed, and I'll follow up with a new release after my holiday, when I hopefully have a moment to add WASI support to the TS runtime too.

Cheers!

**Pull request merged**
